yos/list.go

```diff
@@ -0,0 +1,85 @@
+package yos
+
+import (
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// A FilePathInfo describes a file's path and stat.
+type FilePathInfo struct {
+	Path string
+	Info os.FileInfo
+}
+
+// listCondEntries returns a list of conditional directory entries.
+func listCondEntries(root string, cond func(os.FileInfo) (bool, error)) (entries []*FilePathInfo, err error) {
+	err = filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if root == path {
+			return nil
+		}
+		var ok bool
+		if ok, err = cond(info); ok {
+			entries = append(entries, &FilePathInfo{
+				Path: path,
+				Info: info,
+			})
+		}
+		return err
+	})
+	return
+}
+
+// ListAll returns a list of all directory entries in the given directory in lexical order.
+// It searches recursively, but symbolic links will be not be followed.
+func ListAll(root string) (entries []*FilePathInfo, err error) {
+	return listCondEntries(root, func(info os.FileInfo) (bool, error) { return true, nil })
+}
+
+// ListFile returns a list of file directory entries in the given directory in lexical order.
+// It searches recursively, but symbolic links will be not be followed.
+func ListFile(root string) (entries []*FilePathInfo, err error) {
+	return listCondEntries(root, func(info os.FileInfo) (bool, error) { return !info.IsDir(), nil })
+}
+
+// ListDir returns a list of nested directory entries in the given directory in lexical order.
+// It searches recursively, but symbolic links will be not be followed.
+func ListDir(root string) (entries []*FilePathInfo, err error) {
+	return listCondEntries(root, func(info os.FileInfo) (bool, error) { return info.IsDir(), nil })
+}
+
+const (
+	// The flags are used by the ListMatch methods.
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [comment on exported const ListRecursive should be of the form "ListRecursive ..."](https://app.codacy.com/app/an9an63/gut/pullRequest?prid=4666961)